src/Compilers/CSharp/Portable/Errors/LazyObsoleteDiagnosticInfo.cs

```diff
@@ -54,5 +55,25 @@ internal override DiagnosticInfo GetResolvedInfo()
 
             return _lazyActualObsoleteDiagnostic;
         }
+
+        public override bool Equals(object obj)
+        {
+            return obj is LazyObsoleteDiagnosticInfo lod &&
+                this._lazyActualObsoleteDiagnostic.Equals(lod._lazyActualObsoleteDiagnostic) &&
```

Comment:
>_lazyActualObsoleteDiagnostic [](start = 21, length = 29)

We shouldn't be accessing ```_lazyActualObsoleteDiagnostic``` without making sure it has been initialized properly. I think we have two options here either we completely delegate equality to the result of ```GetResolvedInfo```, or we exclude ```_lazyActualObsoleteDiagnostic``` from the equality completely if we don't want the resolution to be triggered by the equality.